---
title: "Figure 1"
output: 
  html_document: 
    toc: yes
author: 'Pietro Demela, Nicola Pirastu, Blagoje Soskic'
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = TRUE,
	warning = TRUE
)
knitr::opts_knit$set(root.dir =  "/project/aid_sharing/AID_sharing") #set up the project directory
```

## Figure 1A

### Libraries

```{r}
library(corrplot)
```


### Correlation matrix

```{r}
#load the ldsc output
ldsc_ok <- readRDS('outputs/rev_1/02_munge_ldsc_sumstats/ldsc_output_rev1.RDS')

#rename the columns with uppercase names in the right order
colnames(ldsc_ok$S_Stand)
names_plot = c('T1D','CD', 'UC', 
               'PSC', 'JIA', 'SLE', 
               'RA',  'Ast', 'Ecz')


#rename columns and rows
rownames(ldsc_ok$S_Stand) <- names_plot
colnames(ldsc_ok$S_Stand) <- names_plot

#reorder
plot <- ldsc_ok$S_Stand[c('CD', 'UC','PSC', 'JIA', 'SLE','RA','T1D', 'Ast', 'Ecz'),c('CD', 'UC','PSC', 'JIA', 'SLE','RA','T1D', 'Ast' , 'Ecz')]

#save
pdf('outputs/rev_1/Figures/Fig_1/1A_correlation_matrix.pdf', height = 14, width = 14)
corrplot(plot, 
         order = 'original',
         addCoef.col = 'black', 
         method = 'square',
         type = 'upper', 
         is.corr = T,
         tl.col='black',
         outline=F,
         number.cex= 1.5,
         tl.cex=1.5,
         cl.cex=1.5,
         tl.srt= 45,
         addgrid.col='grey', 
         title= 'LDSC genetic correlation matrix of autoimmune diseases', mar=c(0,0,1,0)
)

dev.off()

#plot
corrplot(plot, 
         order = 'original',
         addCoef.col = 'black', 
         method = 'square',
         type = 'upper', 
         is.corr = T,
         tl.col='black',
         outline=F,
         number.cex= 1,
         tl.cex=1.5,
         cl.cex=1.5,
         tl.srt= 45,
         addgrid.col='grey', 
         title= 'LDSC genetic correlation matrix of autoimmune diseases', mar=c(0,0,1,0)
)
```


## Figure 1B

### Libraries

```{r message=FALSE, warning=FALSE}
library(data.table)
library(GenomicSEM)
library(Matrix)
library(tidyr)
library(dplyr)
library(corrplot)
library(qgraph)
library(semPlot)
library(lavaan)
library(RColorBrewer)
library(gridExtra)
```

### Function for SEM pathways

Function that allows to plot the output of usermodel with pathview
```{r}

semPlotModel_GSEM=function(gsem.object=GWISoutput , est.label="STD_All"){ 
  require(semPlot)
  object=gsem.object$results
  object$free=0
  numb=1:length(which(object$op!="~~"))
  object$free[which(object$op!="~~")]=numb
  varNames <- lavaanNames(object, type = "ov")
  factNames <- lavaanNames(object, type = "lv")
  factNames <- factNames[!factNames %in% varNames]
  n <- length(varNames)
  k <- length(factNames)
  if (is.null(object$label)) 
    object$label <- rep("", nrow(object))
  semModel <- new("semPlotModel")
  object$est <- object[,est.label]
  if (is.null(object$group)) 
    object$group <- ""
  semModel@Pars <- data.frame(label = object$label, lhs = ifelse(object$op == 
                                                                   "~" | object$op == "~1", object$rhs, object$lhs), edge = "--", 
                              rhs = ifelse(object$op == "~" | object$op == "~1", object$lhs, 
                                           object$rhs), est = object$est, std = NA, group = object$group, 
                              fixed = object$free==0, par = object$free, stringsAsFactors = FALSE)
  semModel@Pars$edge[object$op == "~~"] <- "<->"
  semModel@Pars$edge[object$op == "~*~"] <- "<->"
  semModel@Pars$edge[object$op == "~"] <- "~>"
  semModel@Pars$edge[object$op == "=~"] <- "->"
  semModel@Pars$edge[object$op == "~1"] <- "int"
  semModel@Pars$edge[grepl("\\|", object$op)] <- "|"
  semModel@Thresholds <- semModel@Pars[grepl("\\|", semModel@Pars$edge), 
                                       -(3:4)]
  semModel@Pars <- semModel@Pars[!object$op %in% c(":=", "<", 
                                                   ">", "==", "|", "<", ">"), ]
  semModel@Vars <- data.frame(name = c(varNames, factNames), 
                              manifest = c(varNames, factNames) %in% varNames, exogenous = NA, 
                              stringsAsFactors = FALSE)
  semModel@ObsCovs <- list()
  semModel@ImpCovs <- list()
  semModel@Computed <- FALSE
  semModel@Original <- list(object)
  return(semModel)
  
}



```


### Path model

```{r message=TRUE, warning=TRUE, paged.print=FALSE}

#load the ldsc output
ldsc_ok <- readRDS('outputs/rev_1/02_munge_ldsc_sumstats/ldsc_output_rev1.RDS')


# the factor model 
aid_model <-'F1 =~ NA*crohn + uc  + psc  
F2 =~ NA*jia + sle + ra + t1d 
F3 =~ NA*asthma + derma 

F1~~F2
F1~~F3
F2~~F3
F1 ~~ 1*F1
F2 ~~ 1*F2
F3 ~~ 1*F3
derma~~a*derma
a>0.001
'

aid_factor <-usermodel(ldsc_ok, estimation = "DWLS", model = aid_model, CFIcalc = TRUE, std.lv = TRUE, imp_cov = FALSE)
signif(aid_factor$modelfit,2)


#save
pdf('outputs/rev_1/Figures/Fig_1/1B_three_factor_model.pdf', height = 20, width = 18)
semPaths(semPlotModel_GSEM(aid_factor), 
         what = 'path' , 
         whatLabels= 'est',
         residuals = T, 
         sizeMan = 8, 
         sizeInt = 10,
         label.cex=1, 
         theme="colorblind", 
         rotation = 4, 
         layout = "tree2", 
         sizeLat = 12,
         edge.color = "black",
         edge.label.cex = 1,
         curve = 2,
         nodeLabels = c('CD', 'UC', 'PSC', 'JIA', 
                        'SLE', 'RA','T1D' ,'AST', 'ECZ' , 'F1', 'F2', 'F3' ), 
         label.cex = 1.5, 
         color = list( lat=brewer.pal(12, 'Paired')[c(1,3,5)]),
         groups=list(c("crohn","uc","psc"),c("jia","sle","ra_eu","t1d"), c( "asthma" ,"derma")),
         height = 18, width = 18, 
         edge.label.position=0.5,
         asize=3,
         esize=1,
         
)
dev.off()


#plot 
semPaths(semPlotModel_GSEM(aid_factor), 
         what = 'path' , 
         whatLabels= 'est',
         residuals = T, 
         sizeMan = 8, 
         sizeInt = 10,
         label.cex=1, 
         theme="colorblind", 
         rotation = 4, 
         layout = "tree2", 
         sizeLat = 12,
         edge.color = "black",
         edge.label.cex = 1,
         curve = 2,
         nodeLabels = c('CD', 'UC', 'PSC', 'JIA', 
                        'SLE', 'RA','T1D' ,'AST', 'ECZ' , 'F1', 'F2', 'F3' ), 
         label.cex = 1.5, 
         color = list( lat=brewer.pal(12, 'Paired')[c(1,3,5)]),
         groups=list(c("crohn","uc","psc"),c("jia","sle","ra_eu","t1d"), c( "asthma" ,"derma")),
         height = 18, width = 18, 
         edge.label.position=0.5,
         asize=3,
         esize=1,
         
)
```


## Figure 1C

### Libraries 

```{r message=FALSE, warning=FALSE, paged.print=TRUE}
library(data.table)
library(qqman)
library(dplyr)
library(CMplot)
library(rcartocolor)
library(RColorBrewer)
```

### Man plot

```{r message=TRUE, warning=TRUE, paged.print=FALSE}

#load the summary statistics
gw <- list()

for(i in c(1:3)){
      tt <- c('f1', 'f2', 'f3')[i]
      a1 <- fread(paste0('outputs/rev_1/03_gwas_estimation/summarystats_',tt,'_.txt' ),
                        data.table = F)
      a1 <-  select(a1, c('SNP', 'P','CHR','BP')) 
      colnames(a1)[2] <- tt
      gw[[tt]] <- a1
}

#create a dataframe with the pvalues for all the three factors (named f1, f2, f3)
circular <- purrr::reduce(gw, dplyr::left_join, by = 'SNP')
circular <- select(circular,-c('CHR.x','BP.x' ,'CHR.y' ,  'BP.y' ))
circular <- circular[,c('SNP','CHR', 'BP', 'f1', 'f2', 'f3')]
head(circular)


#load the factor loci table to highlight the lead SNPs with a rhomboid
factor_loci <- fread('outputs/rev_1/04_genomic_regions_factors/regions_factors.txt', data.table = F) 

#v3

matrix(c(rcartocolor::carto_pal(3, 'Safe')[1],rcartocolor::carto_pal(3, 'Safe')[2],rcartocolor::carto_pal(3, 'Safe')[3],
         RColorBrewer::brewer.pal(4, 'Blues')[4], RColorBrewer::brewer.pal(3, 'Reds')[3], RColorBrewer::brewer.pal(4, 'YlOrRd')[2]),
       3,2,byrow=T)

#this function will save a png in the home directory
  CMplot(circular, 
         plot.type="m",
         multracks=TRUE,
         threshold=c(5e-8),
         threshold.lty=c(1,2), 
         threshold.lwd=c(1,1), 
         threshold.col=c("black","grey"), 
         amplify=F,
         #signal.col=list(brewer.pal(12, 'Paired')[c(1)], brewer.pal(12, 'Paired')[3], brewer.pal(12, 'Paired')[c(5)]),
         #signal.pch=c(1),
         signal.cex=0.7,
         col=matrix(c(brewer.pal(12, 'Paired')[c(1,3,5)], brewer.pal(12, 'Paired')[c(2,4,6)]),
                    3,2,byrow=F),
         highlight=list(f1=factor_loci[which(factor_loci$trait=='f1'), ]$SNP, f2=factor_loci[which(factor_loci$trait=='f2'), ]$SNP, f3=factor_loci[which(factor_loci$trait=='f3'), ]$SNP),
         highlight.col='black',
         highlight.pch = 18, 
         highlight.cex=1.6,
         file=c("jpg"),
         memo="v3",
         dpi=300,
         file.output=T,
         verbose=TRUE, 
         LOG10=T,
         chr.border=F,
         cex.lab=2,
         width=50, 
         height=10)
  
system('mv Mul* /project/aid_sharing/AID_sharing/outputs/rev_1/Figures/Fig_1/')

```


## Figure 1D

### Libraries

```{r}
library(data.table)
library(stringr)
library(GenomicRanges)
library(dplyr)
library(ChIPpeakAnno)
library(RColorBrewer)
library(ComplexHeatmap)
```


###  Function to use the ChIPpeakAnno::findOverlapsOfPeaks() for plotting upset plots

This function is for obtaining a combination matrix for plotting and UpSet plot from Granges names. It uses the names of the peaks from the results of findOverlapsOfPeaks. 
It only works in the case of a three-element list. (Tested only with 2 and 3 groups)

```{r}

lists_forupset <- function(peaks, traits){ 
  #allocate lists
  require(ComplexHeatmap)
  require(ChIPpeakAnno)
  require(stringr)
  
  ovl_f <- ChIPpeakAnno::findOverlapsOfPeaks(peaks)
  
  un <- list()
  final <- list(list())
  sh <- list()
  #create the lists and put it the elements that are unique for each trait
  for( i in 1:length(traits)){
    tt<-traits[i]
    filt <- str_starts(names(ovl_f$uniquePeaks@ranges),tt, negate = FALSE)
    un[[tt]] <- names(ovl_f$uniquePeaks@ranges)[filt]
  }
  
  #put into the lists the element that are shared among the lists
  for(k in 1:length(ovl_f$mergedPeaks$peakNames)){
    
    check_in <- traits %in% substring(ovl_f$mergedPeaks$peakNames[[k]], first = 1, last = 2) #check if there is f1, f2 or f3
    
    for(u in 1:length(check_in)){
      tt<-traits[u]
      if(check_in[u]==T){
        un[[tt]][[length(un[[tt]])+1]] <- paste0(ovl_f$mergedPeaks$peakNames[[k]], collapse = '-')
      }
      
    }
  }
  
  #generate the output
 
  cm <- make_comb_mat(un, mode = 'distinct')
  return(cm)
  
}
```


### Prepare the GenomiRanges objects of the factor genomic regions
One list with three elements. Each element is the ranges of the genomic regions for that factor. 

```{r}
#load the factors genomic regions
regions_factors <- fread('outputs/rev_1/04_genomic_regions_factors/regions_factors.txt', data.table = F)

f_ranges <- list()

for(i in c('f1', 'f2', 'f3')){

    f_ranges[[i]] <- GRanges(seqnames = regions_factors[which(regions_factors$trait==i),]$chr,   IRanges(as.numeric(regions_factors[which(regions_factors$trait==i),]$start),as.numeric(regions_factors[which(regions_factors$trait==i),]$end)))

}
```

### Plot Upset plot

```{r message=TRUE, warning=TRUE, paged.print=FALSE}
cm <- lists_forupset(f_ranges, traits = c('f1', 'f2', 'f3')) 

#save
pdf(width = 6, height = 5, file = 'outputs/rev_1/Figures/Fig_1/1D_upset_f1_f2_f3_loci.pdf')
#plot
UpSet(cm, set_order = c("f1", "f2", "f3"), comb_order = order(c(1,3,4,2,7,6,5), decreasing = F),
      comb_col = c(brewer.pal(8, 'Set2')[c(7,6,6,6)],brewer.pal(12, 'Paired')[c(2,4,6)]),
      top_annotation = upset_top_annotation(cm, add_numbers = TRUE, height = unit(6, "cm")),
      left_annotation = upset_left_annotation(cm, add_numbers = TRUE, width = unit(3,'cm'), gp = gpar(fill = brewer.pal(5, 'Greys')[1])  ),
      row_title = "Factor", 
      column_title = " Physical intersection of factor loci"
)
dev.off()

#plot
UpSet(cm, set_order = c("f1", "f2", "f3"), comb_order = order(c(1,3,4,2,7,6,5), decreasing = F),
      comb_col = c(brewer.pal(8, 'Set2')[c(7,6,6,6)],brewer.pal(12, 'Paired')[c(2,4,6)]),
      top_annotation = upset_top_annotation(cm, add_numbers = TRUE, height = unit(6, "cm")),
      left_annotation = upset_left_annotation(cm, add_numbers = TRUE, width = unit(3,'cm'), gp = gpar(fill = brewer.pal(5, 'Greys')[1])  ),
      row_title = "Factor", 
      column_title = " Physical intersection of factor loci"
)


```
#### Upset vs venndiagram

Make sure the venndiagram produces the same reuslts of the Upset. 

```{r}
ovl_f <- ChIPpeakAnno::findOverlapsOfPeaks(f_ranges , connectedPeaks = 'keepAll')
makeVennDiagram(ovl_f)
```

## Figure 1E
### Libraries

```{r}
library(data.table)
library(stringr)
library(ComplexHeatmap)
library(RColorBrewer)
library(circlize)
```

### Calculate z scores for the per-factor unique lead SNPs

```{r message=TRUE, warning=TRUE, paged.print=FALSE}
#load the regions with the index of overlapp
factor_loci <- fread('outputs/rev_1/04_genomic_regions_factors/regions_factors_with_index.txt', data.table = F)

#load the summary statistic of the factors and retrieve only the lead SNPs of the genomic regions (all of them for each of the factor)
f_list <- list()
f_snps <- list()

  for(i in 1:3){
    
          tt <- c('f1','f2', 'f3')[i]
          f_list[[tt]] <- fread(paste0('outputs/rev_1/03_gwas_estimation/summarystats_',tt,'_.txt'), data.table = F)
          f_snps[[tt]] <- f_list[[tt]][which(f_list[[tt]]$SNP %in% factor_loci$SNP), ]
  }

#add 'locus' to the index of the overlapping groups
factor_loci$pan_locus <- paste0('locus', factor_loci$pan_locus)

#put all the summary stats togehter, the rownames will indicate to which factor the summary stats belong. 
all_f_s <- do.call(rbind, f_snps)
head(all_f_s)
table(str_sub(rownames(all_f_s), start = 1, end = 2))

#calculate the z score of each SNP
all_f_s$z_score <- all_f_s$Beta/all_f_s$SE


#create a dataframe where each row is one of the enomic regions, and 5 columns, the z score for each factor (F1, F2, F3), 
#a trait column that will state if that SNP is significant for that factor
#a leads column which will contain the SNPs that are lead for that genomic region
to_plot <- data.frame(row.names = unique(factor_loci$pan_locus_name), 
                      F1=rep(NA, length(unique(factor_loci$pan_locus_name))),  
                      F2=rep(NA, length(unique(factor_loci$pan_locus_name))),
                      F3=rep(NA, length(unique(factor_loci$pan_locus_name))), 
                      traits=rep(NA, length(unique(factor_loci$pan_locus_name))),
                      leads=rep(NA, length(unique(factor_loci$pan_locus_name))))
head(to_plot)

#populate the dataframe; for each row check which factor gwas is significant for that locus and return the lead SNP
#if more than one factor is significant for that genomic regions, report both and paste the leads and traits names.
  for(i in 1:nrow(to_plot)){
    
        tt <- rownames(to_plot)[i]
        to_plot[tt, ]$traits <- paste0(factor_loci[which(factor_loci$pan_locus_name== tt),]$trait, collapse = '-')
        to_plot[tt, ]$leads <- paste0(factor_loci[which(factor_loci$pan_locus_name== tt), ]$SNP, collapse = '-')
    
  }

head(to_plot)

mean(to_plot[which(to_plot$traits %in% c('f1', 'f2', 'f3')),]$leads %in% all_f_s$SNP) #all SNP are present in the factor specifics group



#exclude the genomic regions present in more than one factor
to_plot_f <- to_plot[which(to_plot$traits %in% c('f1', 'f2', 'f3')),]

dim(to_plot_f) #133 


#add the zeta scores
for(i in 1:3){
  
     #for each factor
    tt <- c('f1','f2', 'f3')[i]
    
    for(k in 1:nrow(to_plot_f)){
      #retrieve the zeta score from all_f_s by using the row name and the SNP name
          snp <- to_plot_f[k,]$leads
          to_plot_f[which(to_plot_f$leads==snp), toupper(tt)]  <-   all_f_s[which(str_starts(rownames(all_f_s),tt) & all_f_s$SNP==snp),]$z_score
          
    }
    
}

head(to_plot_f)

```

### HeatMap z-scores

```{r}
#save the heatmap
pdf('outputs/rev_1/Figures/Fig_1/1E_heatmap_complete_absolute_z_scores_specifc_factors.pdf', height = 16, width = 6)
Heatmap(as.matrix(abs(to_plot_f[,1:3])), #transform it into absolute values
        col = colorRamp2(c(seq(1,13, length.out=9)), c(RColorBrewer::brewer.pal(9,'Purples'))), 
        column_title = "Heatmap of absolute z scores" , 
        cluster_rows = T,
        cluster_columns = F,
        cluster_column_slices = F,
        cluster_row_slices =  F,
        show_row_dend = F,
        column_split =  c('F1', 'F2', 'F3'),
        row_split = factor(toupper(to_plot_f[,'traits']), levels = c('F1', 'F2', 'F3')),
        column_gap = unit(3, "mm"),
        row_gap = unit(5, "mm") ,
        border = T,
        row_names_gp = gpar( fontsize=0), 
        # row_title='Lead SNP',
        heatmap_legend_param =list(title = "Absosule z score"))
dev.off()

#plot
Heatmap(as.matrix(abs(to_plot_f[,1:3])),
        col = colorRamp2(c(seq(1,13, length.out=9)), c(RColorBrewer::brewer.pal(9,'Purples'))), 
        column_title = "Heatmap of absolute z scores" , 
        cluster_rows = T,
        cluster_columns = F,
        cluster_column_slices = F,
        cluster_row_slices =  F,
        show_row_dend = F,
        column_split =  c('F1', 'F2', 'F3'),
        row_split = factor(toupper(to_plot_f[,'traits']), levels = c('F1', 'F2', 'F3')),
        column_gap = unit(3, "mm"),
        row_gap = unit(5, "mm") ,
        border = T,
        row_names_gp = gpar( fontsize=0), 
        # row_title='Lead SNP',
        heatmap_legend_param =list(title = "Absosule z score"))

```


```{r}
sessionInfo()
```









