---
title: "02_munge_ldsc_sumstats"
author: "Pietro Demela"
date: "21/12/2022"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = TRUE,
	warning = TRUE
)
knitr::opts_knit$set(root.dir =  "/project/aid_sharing/AID_sharing") #set up the project directory
```


## LD score regression on auotoimmunity GWAS summary statistics
In this script the selected summary stats will be used to run ldsc, munge and sum stats function. 
Importat: the have to be in the same order in the function

```{r libraries, message=TRUE, warning=TRUE, paged.print=TRUE}
library(data.table)
library(GenomicSEM)
library(tidyr)
library(dplyr)
library(corrplot)
library(RColorBrewer)
```

## Function for effective sample size calculations
A function for doing it rapidly from csv files. The csv files generated by me, from info in the papers
```{r calculate_prevalence, message=TRUE, warning=TRUE, paged.print=TRUE}

calculate_prevalence <- function(path_of_csv){
  #read the file and remove the NA columns and rows
  preva_csv <- read.csv(file = path_of_csv , header = T, sep = ';')
  preva_csv <-  preva_csv[, c('cases', 'controls')]
  preva_csv_noNA <- preva_csv[complete.cases(preva_csv), ]
  
  #calculated effective prevalence 
  #calculate sample prevalence for each cohort
  preva_csv_noNA$v <-preva_csv_noNA$cases/(preva_csv_noNA$cases+preva_csv_noNA$controls)
  #calculate cohort specific effective sample size
  preva_csv_noNA$EffN<-4*preva_csv_noNA$v*(1-preva_csv_noNA$v)*(preva_csv_noNA$cases+preva_csv_noNA$controls)
  #calculate sum of effective sample size: 
  eff_sample_size <- round(sum(preva_csv_noNA$EffN), 3)
  
  #print number of cases and controls, and effective sample size
  cat(paste0( 'Number of cases  ' , sum(preva_csv_noNA$cases), '\n', 
              'Number of contros  ', sum(preva_csv_noNA$controls), '\n', 
              'Effective sample size  ', round(sum(preva_csv_noNA$EffN), 3)
  ))
  
  #output the prevalenc when assigned to a variable 
  invisible(round(eff_sample_size, 3))
}

```



## Munge

```{r}
file_names <- c('outputs/rev_1/01_prepare/ready_format/t1d_chiou-2021.txt', 
                'outputs/rev_1/01_prepare/ready_format/cd_build37_delange-2017.txt', 
                'outputs/rev_1/01_prepare/ready_format/uc_build37_delange-2017.txt',
                ###
                'outputs/rev_1/01_prepare/ready_format/psc_ji-2016.txt',
                'outputs/rev_1/01_prepare/ready_format/jia_beta_lopezisac-2020.txt',
                'outputs/rev_1/01_prepare/ready_format/sle_beta_bentham-2015.txt',
                ###
                'outputs/rev_1/01_prepare/ready_format/ra_eu_okada-2014.txt',
                'outputs/rev_1/01_prepare/ready_format/asthma_han-2020.txt',
                'outputs/rev_1/01_prepare/ready_format/derma_sliz-2021.txt'
                )

traits <- c('t1d','crohn', 'uc', 
            'psc', 'jia', 'sle',
            'ra',  'asthma', 'derma')

t1d_p <- calculate_prevalence('Prevalences/CSV_prevalences/t1d_chiou-2021.csv')
crohn_p <- calculate_prevalence('Prevalences/CSV_prevalences/crohn_delange-2017.csv')
uc_p <- calculate_prevalence('Prevalences/CSV_prevalences/uc_delange-2017.csv')
psc_p <- 14890
jia_p <- 12501
sle_p <- calculate_prevalence('Prevalences/CSV_prevalences/sle_benthman-2015.csv')
ra_p <- calculate_prevalence('Prevalences/CSV_prevalences/ra_okada-2014_only-eu.csv')
asthma_p <- 64538 + 329321
derma_p <- calculate_prevalence('Prevalences/CSV_prevalences/derma_sliz-2021.csv')

munge(files = file_names, hm3 = 'SNP/w_hm3.snplist', 
      N = c(t1d_p, crohn_p, uc_p,
            psc_p, jia_p, sle_p, 
            ra_p, asthma_p, derma_p), 
      trait.names = traits,
      parallel = T,
      cores = 9)
```

Move log files and summary stats to their folders
```{r move munged and log, message=TRUE, warning=TRUE, paged.print=TRUE}

system('mv *gz /project/aid_sharing/AID_sharing/outputs/rev_1/02_munge_ldsc_sumstats/munged')
system('mv *log /project/aid_sharing/AID_sharing/outputs/rev_1/02_munge_ldsc_sumstats/log_munge')

```

## ldsc

```{r ldsc, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
munged_files <- c('outputs/rev_1/02_munge_ldsc_sumstats/munged/t1d.sumstats.gz',
                  'outputs/rev_1/02_munge_ldsc_sumstats/munged/crohn.sumstats.gz',
                  'outputs/rev_1/02_munge_ldsc_sumstats/munged/uc.sumstats.gz',
                  ###
                  'outputs/rev_1/02_munge_ldsc_sumstats/munged/psc.sumstats.gz',
                  'outputs/rev_1/02_munge_ldsc_sumstats/munged/jia.sumstats.gz', 
                  'outputs/rev_1/02_munge_ldsc_sumstats/munged/sle.sumstats.gz', 
                  ###
                  'outputs/rev_1/02_munge_ldsc_sumstats/munged/ra.sumstats.gz', 
                  'outputs/rev_1/02_munge_ldsc_sumstats/munged/asthma.sumstats.gz', 
                  'outputs/rev_1/02_munge_ldsc_sumstats/munged/derma.sumstats.gz')

names = c('t1d','crohn', 'uc', 
          'psc', 'jia', 'sle', 
          'ra',  'asthma', 'derma')



ldsc_output <- ldsc(traits = munged_files, 
                      sample.prev = c(.5, .5, .5, 
                                      0.1928, 0.2643, .5, 
                                      .5, ((64538)/(64538 + 329321)), .5),  
                     population.prev = c(0.095, 0.001, 0.0003, 
                                         0.000050, 0.000447, 0.000500,  
                                         0.0046, 0.035700, 0.20 ), 
                     trait.names = names,
                     ld = "ldscores/eur_w_ld_chr",
                     wld= "ldscores/eur_w_ld_chr", stand = T)

saveRDS(ldsc_output, 'outputs/rev_1/02_munge_ldsc_sumstats/ldsc_output_rev1.RDS')

#move the log file
system('mv *log /project/aid_sharing/AID_sharing/outputs/rev_1/02_munge_ldsc_sumstats/')
```


Plot the correlation matrix
```{r message=TRUE, warning=TRUE, paged.print=TRUE}
ldsc_ok <- readRDS('outputs/rev_1/02_munge_ldsc_sumstats/ldsc_output_rev1.RDS')

names_plot = c('Type 1 Diabetes','Crohn\'s disease', 'Ulcerative colitis', 
               'Primary sclerosing cholangitis ', 'Juvenile idiopathic arthritis', 'Systemic lupus erythematosus', 
               'Rheumatoid arthritis',  'Asthma', 'Atopic dermatitis')

rownames(ldsc_ok$S_Stand) <- names_plot
colnames(ldsc_ok$S_Stand) <- names_plot

corrplot(ldsc_ok$S_Stand, order = 'hclust',addCoef.col = 'black', method = 'square',type = 'upper', is.corr = T,
        tl.col='black', outline=F,
        number.cex= 1,
         tl.cex=0.5,
        cl.cex=1.5,
         tl.srt= 45,
         addgrid.col='grey',
        title= 'LDSC genetic correlation matrix of autoimmune diseases', mar=c(0,0,1,0)
         )
```



## Factor models

### function for plotting sempaths
```{r message=TRUE, warning=TRUE}
library(semPlot)
library(lavaan)
semPlotModel_GSEM=function(gsem.object=GWISoutput , est.label="STD_All"){ 
  require(semPlot)
  object=gsem.object$results
  object$free=0
  numb=1:length(which(object$op!="~~"))
  object$free[which(object$op!="~~")]=numb
  varNames <- lavaanNames(object, type = "ov")
  factNames <- lavaanNames(object, type = "lv")
  factNames <- factNames[!factNames %in% varNames]
  n <- length(varNames)
  k <- length(factNames)
  if (is.null(object$label)) 
    object$label <- rep("", nrow(object))
  semModel <- new("semPlotModel")
  object$est <- object[,est.label]
  if (is.null(object$group)) 
    object$group <- ""
  semModel@Pars <- data.frame(label = object$label, lhs = ifelse(object$op == 
                                                                   "~" | object$op == "~1", object$rhs, object$lhs), edge = "--", 
                              rhs = ifelse(object$op == "~" | object$op == "~1", object$lhs, 
                                           object$rhs), est = object$est, std = NA, group = object$group, 
                              fixed = object$free==0, par = object$free, stringsAsFactors = FALSE)
  semModel@Pars$edge[object$op == "~~"] <- "<->"
  semModel@Pars$edge[object$op == "~*~"] <- "<->"
  semModel@Pars$edge[object$op == "~"] <- "~>"
  semModel@Pars$edge[object$op == "=~"] <- "->"
  semModel@Pars$edge[object$op == "~1"] <- "int"
  semModel@Pars$edge[grepl("\\|", object$op)] <- "|"
  semModel@Thresholds <- semModel@Pars[grepl("\\|", semModel@Pars$edge), 
                                       -(3:4)]
  semModel@Pars <- semModel@Pars[!object$op %in% c(":=", "<", 
                                                   ">", "==", "|", "<", ">"), ]
  semModel@Vars <- data.frame(name = c(varNames, factNames), 
                              manifest = c(varNames, factNames) %in% varNames, exogenous = NA, 
                              stringsAsFactors = FALSE)
  semModel@ObsCovs <- list()
  semModel@ImpCovs <- list()
  semModel@Computed <- FALSE
  semModel@Original <- list(object)
  return(semModel)
  
}

```

### EFA models

```{r EFA models, echo=TRUE, message=TRUE, warning=TRUE, paged.print=TRUE}
require(Matrix)
Ssmooth<-as.matrix((nearPD(ldsc_ok$S, corr = FALSE))$mat)

#run EFA with promax rotation and 2 factors using the factanal function in the stats package
require(stats)
EFA_1 <-factanal(covmat = Ssmooth, factors = 1, rotation = "promax")
EFA_1
EFA_2 <- factanal(covmat = Ssmooth, factors = 2, rotation = "promax")
EFA_2
EFA_3 <- factanal(covmat = Ssmooth, factors = 3, rotation = "promax")
EFA_3 
EFA_4 <- factanal(covmat = Ssmooth, factors = 4, rotation = "promax")
EFA_4
EFA_5 <- factanal(covmat = Ssmooth, factors = 5, rotation = "promax")
EFA_5

#print the loadings
EFA_1$loadings
EFA_2$loadings
EFA_3$loadings
EFA_4$loadings
EFA_5$loadings




```

### One factot model

```{r message=TRUE, warning=TRUE, paged.print=FALSE}
one_f_model <-'F1 =~ NA*crohn + uc  + psc  +jia + sle + ra + t1d +asthma + derma 

F1 ~~ 1*F1

'

one_factor <-usermodel(ldsc_ok, estimation = "DWLS", model = one_f_model, CFIcalc = TRUE, std.lv = TRUE, imp_cov = FALSE)
one_factor

#plot the model
semPaths(semPlotModel_GSEM(one_factor), what = 'path' , 
         whatLabels= 'est',
         residuals = T, 
         sizeMan = 7, 
         sizeInt = 2,
         label.cex=1, 
         theme="colorblind", 
         rotation = 4, 
         layout = "tree2", 
         sizeLat = 7,
         edge.color = "black",
         edge.label.cex = 1,
         curve = 2,
         nodeLabels = c('CD', 'UC', 'PSC', 'JIA', 
                        'SLE', 'RA','T1D' ,'AST', 'ECZ' , 'F'), 
         label.cex = 1.5,
         height = 10, width = 16, 
         edge.label.position=0.5,
         asize=3,
         esize=1)

```


### Two factor model

```{r message=TRUE, warning=TRUE, paged.print=FALSE}
two_f_model <-'F1 =~ NA*crohn + uc  + psc  +jia + sle + ra + t1d 
F2 =~ NA*asthma + derma 

F1~~F2

F1 ~~ 1*F1
F2 ~~ 1*F2

derma~~a*derma
a>0.001
'

two_factor <-usermodel(ldsc_ok, estimation = "DWLS", model = two_f_model, CFIcalc = TRUE, std.lv = TRUE, imp_cov = FALSE)
two_factor

semPaths(semPlotModel_GSEM(two_factor), what = 'path' , 
         whatLabels= 'est',
         residuals = T, 
         sizeMan = 7, 
         sizeInt = 2,
         label.cex=1, 
         theme="colorblind", 
         rotation = 4, 
         layout = "tree2", 
         sizeLat = 7,
         edge.color = "black",
         edge.label.cex = 1,
         curve = 2,
        nodeLabels = c('CD', 'UC', 'PSC', 'JIA', 
                       'SLE', 'RA','T1D' ,'AST', 'ECZ' , 'F1', 'F2'), 
         height = 10, width = 16, 
         edge.label.position=0.5,
         asize=3,
         esize=1)


```



### Three factor model

```{r message=TRUE, warning=TRUE, paged.print=FALSE}

aid_model <-'F1 =~ NA*crohn + uc  + psc  
F2 =~ NA*jia + sle + ra + t1d 
F3 =~ NA*asthma + derma 

F1~~F2
F1~~F3
F2~~F3
F1 ~~ 1*F1
F2 ~~ 1*F2
F3 ~~ 1*F3
derma~~a*derma
a>0.001
'


#run the model
aid_factor <-usermodel(ldsc_ok, estimation = "DWLS", model = aid_model, CFIcalc = TRUE, std.lv = TRUE, imp_cov = FALSE)
aid_factor

semPaths(semPlotModel_GSEM(aid_factor), 
         what = 'path' , 
         whatLabels= 'est',
         residuals = T, 
         sizeMan = 8, 
         sizeInt = 10,
         label.cex=1, 
         theme="colorblind", 
         rotation = 4, 
         layout = "tree2", 
         sizeLat = 12,
         edge.color = "black",
         edge.label.cex = 1,
         curve = 2,
         nodeLabels = c('CD', 'UC', 'PSC', 'JIA', 
                        'SLE', 'RA','T1D' ,'AST', 'ECZ' , 'F1', 'F2', 'F3' ), 
         label.cex = 1.5, 
         color = list( lat=brewer.pal(12, 'Paired')[c(1,3,5)]),
         groups=list(c("crohn","uc","psc"),c("jia","sle","ra_eu","t1d"), c( "asthma" ,"derma")),
         height = 18, width = 18, 
         edge.label.position=0.5,
         asize=3,
         esize=1,
         
)
```

### Four factor model
```{r message=TRUE, warning=TRUE, paged.print=FALSE}

aid_model_4 <-'F1 =~ NA*crohn + uc 
F2 =~ NA*jia + sle + ra + t1d 
F3 =~ NA*asthma + derma 
F4 =~ NA*psc + uc

F1~~F2
F1~~F3
F2~~F3
F1~~F4
F2~~F4
F3~~F4

F4 ~~ 1*F4
F1 ~~ 1*F1
F2 ~~ 1*F2
F3 ~~ 1*F3
derma~~a*derma
a>0.001
'


#run the model
aid_factor_4 <-usermodel(ldsc_ok, estimation = "DWLS", model = aid_model_4, CFIcalc = TRUE, std.lv = TRUE, imp_cov = FALSE)
aid_factor_4

semPaths(semPlotModel_GSEM(aid_factor_4), what = 'path' , 
         whatLabels= 'est',
         residuals = T, 
         sizeMan = 7, 
         sizeInt = 2,
         label.cex=1, 
         theme="colorblind", 
         rotation = 4, 
         layout = "tree2", 
         sizeLat = 7,
         edge.color = "black",
         edge.label.cex = 1,
         curve = 2,
         #nodeLabels = c('CD', 'UC', 'PSC', 'JIA', 
                       # 'SLE', 'RA','T1D' ,'AST', 'ECZ' , 'F1', 'F2', 'F3', 'F4'), 
         height = 10, width = 16, 
         edge.label.position=0.5,
         asize=3,
         esize=1)


```


## Sumstats 

### Check SE or SE of odds ratio
This function will randomly take 10000 SNPs and will check wheter the reported SE are SE of logistic beta (by calculating a pvalue and comparing it with the reported pvalue)
```{r is_se_logB function, message=TRUE, warning=TRUE, paged.print=TRUE}

is_se_logB <- function(BETA,SE, PVALUE) {
  p_calculated <- 2*pnorm((abs(BETA) / SE),lower.tail = F)
  p_reported <- PVALUE
 
  rand <- sample(1:length(BETA), 10000)
  a <- lm(p_calculated[rand]~p_reported[rand]) #fit a linear model to see if they are perfectly correlated 
  print(summary(a))
  plot(p_reported[rand], p_calculated[rand]) 
  abline(a, col="red") # regression line (y~x)
  abline(0, 1, col='blue', lty= 4 )
}

```


```{r check se or se odds ratio, message=TRUE, warning=TRUE, paged.print=FALSE}
#---- crohn SE -----------------------------------------------------------------
cd <- fread('outputs/rev_1/01_prepare/ready_format/cd_build37_delange-2017.txt', data.table = F)
head(cd) 
is_se_logB(cd$Beta, cd$SE, cd$p) #SE logistic beta (effect column is beta as there are negative values)

#----- uc SE --------------------------------------------------------------------
uc <- fread('outputs/rev_1/01_prepare/ready_format/uc_build37_delange-2017.txt', data.table = F)
head(uc)
is_se_logB(uc$Beta, uc$SE, uc$p) #SE logistic beta (effect column is beta as there are negative values)

#----- psc SE -------------------------------------------------------------------
psc <- fread('outputs/rev_1/01_prepare/ready_format/psc_ji-2016.txt', data.table = F)
head(psc)

is_se_logB(log(psc$OR), psc$SE, psc$p) #SE of logistic beta

#----- jia SE ------------------------------------------------------------------

jia  <- fread('outputs/rev_1/01_prepare/ready_format/jia_beta_lopezisac-2020.txt', data.table = F) 
head(jia)
is_se_logB(jia$Beta, jia$SE, jia$p) #SE of logistic beta

#----- sle SE ------------------------------------------------------------------

sle <- fread('outputs/rev_1/01_prepare/ready_format/sle_beta_bentham-2015.txt', data.table = F)
head(sle)
is_se_logB(BETA = sle$Beta, SE = sle$SE, PVALUE = sle$p) #SE of logistic beta (effect column is beta as there are negative values)

#----- ra eu okada SE -------------------------------------------------------------------

ra <- fread( 'outputs/rev_1/01_prepare/ready_format/ra_eu_okada-2014.txt', data.table = F)
head(ra)
is_se_logB(BETA = ra$Beta, SE = ra$SE, PVALUE = ra$p) #SE of logistic beta

#-----t1d SE----------------------------------------------------------------------

t1d <- fread('outputs/rev_1/01_prepare/ready_format/t1d_chiou-2021.txt', data.table = F)
head(t1d)
is_se_logB(BETA = t1d$Beta, SE = t1d$SE, PVALUE = t1d$p) #SE of logistic BETA


#------asthma 2 SE -------------------------------------------------------------

asthma <- fread('outputs/rev_1/01_prepare/ready_format/asthma_han-2020.txt', data.table = F)
head(asthma)
is_se_logB(log(asthma$OR), asthma$SE, asthma$p) #SE of logistic BETA, the analysis is logfistic regression

#----derma SE ---------------------------------------------------------------
derma <- fread('outputs/rev_1/01_prepare/ready_format/derma_sliz-2021.txt', data.table = F)
head(derma)
is_se_logB(derma$Beta, derma$SE, derma$p) #SE of logistic beta, logistic regression checked on paper

```


```{r sumstats, message=TRUE, warning=TRUE, paged.print=TRUE}

file_names_ok <-c('outputs/rev_1/01_prepare/ready_format/t1d_chiou-2021.txt', 
                  'outputs/rev_1/01_prepare/ready_format/cd_build37_delange-2017.txt', 
                  'outputs/rev_1/01_prepare/ready_format/uc_build37_delange-2017.txt',
                  ###
                  'outputs/rev_1/01_prepare/ready_format/psc_ji-2016.txt',
                  'outputs/rev_1/01_prepare/ready_format/jia_beta_lopezisac-2020.txt',
                  'outputs/rev_1/01_prepare/ready_format/sle_beta_bentham-2015.txt',
                  ###
                  'outputs/rev_1/01_prepare/ready_format/ra_eu_okada-2014.txt',
                  'outputs/rev_1/01_prepare/ready_format/asthma_han-2020.txt',
                  'outputs/rev_1/01_prepare/ready_format/derma_sliz-2021.txt'
                )


aid_sumstats <-sumstats(files = file_names_ok, ref = 'SNP/reference.1000G.maf.0.005.txt.gz',
                                trait.names = c('t1d','crohn', 'uc',
                                                'psc', 'jia', 'sle', 
                                                'ra',  'asthma', 'derma') ,
                                se.logit =  c(T,T,T,T,T,T,T,T,T),  #all case control studies analysed with logistic regression or Odds Ratio reported
                                OLS = c(F,F,F,F,F,F,F,F,F), 
                                linprob = c(F,F,F,F,F,F,F,F,F), 
                                parallel = T, cores = 9,
                                keep.indel= F 
)

fwrite(aid_sumstats, 'outputs/rev_1/02_munge_ldsc_sumstats/aid_sumstats.txt', sep = '\t', col.names = T, row.names = F, quote = F)
system('mv *log /project/aid_sharing/AID_sharing/outputs/rev_1/02_munge_ldsc_sumstats/')


```









