---
title: "06 Genomic regions all traits"
author: "Pietro Demela"
date: "09/01/2023"
output: html_document
---

We define the boundaries of each significant genomic region by identifying all the SNPs with a p-value lower than 1×10-6. We calculated the distance among each consecutive SNPs below
this threshold in the same chromosome; if two SNPs were further than 250 kb apart, then they were defined as belonging to two different genomic regions. We then considered as ‘significant’ all the genomic regions where at least one SNP had a p-value < 5×10-8 (locus.breaker function). This procedure was repeated for all GWAS. Finally, we compared genomic regions between different GWAS and merged those which overlapped, redefining the boundaries as the minimum and maximum genomic position across all overlapping genomic regions (locus_lister function finds the regions that overlap).

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = TRUE,
	warning = TRUE
)
knitr::opts_knit$set(root.dir =  "/project/aid_sharing/AID_sharing") #set up the project directory
```

## Libraries

```{r, libraries}
library(data.table)
library(GenomicRanges)
library(dplyr)
library(ChIPpeakAnno)
library(RColorBrewer)
library(ComplexHeatmap)
library(stringr)
```


## locus.breaker function

This function identifies genome wide significant genomic regions. 
```{r}
locus.breaker <- function (res, p.sig = 5e-08, p.limit = 1e-05, hole.size = 250000, 
  p.label = "P", chr.label = "CHR", pos.label = "BP") 
{
  res = res[order(as.numeric(res[, chr.label]), as.numeric(res[, 
    pos.label])), ]
  res = res[which(res[, p.label] < p.limit), ]
  trait.res = c()
  for (j in 1:22) {
    res.chr = res[which(res[, chr.label] == j), ]
    if (nrow(res.chr) > 1) {
      holes = res.chr[, pos.label][-1] - res.chr[, pos.label][-length(res.chr[, 
        pos.label])]
      gaps = which(holes > hole.size)
      if (length(gaps) > 0) {
        for (k in 1:(length(gaps) + 1)) {
          if (k == 1) {
            res.loc = res.chr[1:(gaps[k]), ]
          }
          else if (k == (length(gaps) + 1)) {
            res.loc = res.chr[(gaps[k - 1] + 1):nrow(res.chr), 
              ]
          }
          else {
            res.loc = res.chr[(gaps[k - 1] + 1):(gaps[k]), 
              ]
          }
          if (min(res.loc[, p.label]) < p.sig) {
            start.pos = min(res.loc[, pos.label], na.rm = T)
            end.pos = max(res.loc[, pos.label], na.rm = T)
            chr = j
            best.snp = res.loc[which.min(res.loc[, p.label]), 
              ]
            line.res = c(chr, start.pos, end.pos, unlist(best.snp))
            trait.res = rbind(trait.res, line.res)
          }
        }
      }
      else {
        res.loc = res.chr
        if (min(res.loc[, p.label]) < p.sig) {
          start.pos = min(res.loc[, pos.label], na.rm = T)
          end.pos = max(res.loc[, pos.label], na.rm = T)
          chr = j
          best.snp = res.loc[which.min(res.loc[, p.label]), 
            ]
          line.res = c(chr, start.pos, end.pos, unlist(best.snp))
          trait.res = rbind(trait.res, line.res)
        }
      }
    }
    else if (nrow(res.chr) == 1) {
      res.loc = res.chr
      if (min(res.loc[, p.label]) < p.sig) {
        start.pos = min(res.loc[, pos.label], na.rm = T)
        end.pos = max(res.loc[, pos.label], na.rm = T)
        chr = j
        best.snp = res.loc[which.min(res.loc[, p.label]), 
          ]
        line.res = c(chr, start.pos, end.pos, unlist(best.snp))
        trait.res = rbind(trait.res, line.res)
      }
    }
  }
  print(trait.res)
  trait.res = as.data.frame(trait.res, stringsAsFactors = FALSE)
  trait.res = trait.res[, -(which(names(trait.res) == chr.label))]
  names(trait.res)[1:3] = c("chr", "start", "end")
  trait.res
}

```

### add log(OR) to asthma and psc

```{r}
asthma <- fread('outputs/rev_1/05_munge/munged/asthma_munged_build37.txt')
asthma$BETA <- log(asthma$OR)
head(asthma)
fwrite(asthma, 'outputs/rev_1/05_munge/munged/asthma_munged_build37.txt', sep = '\t', col.names = T, row.names = F, quote = F)

psc <- fread('outputs/rev_1/05_munge/munged/psc_munged_build37.txt')
psc$BETA <- log(psc$OR)
head(psc)
fwrite(psc, 'outputs/rev_1/05_munge/munged/psc_munged_build37.txt', sep = '\t', col.names = T, row.names = F, quote = F)
```

## locus_lister function

This function identifies significant genomic regions with the locus.breaker function. Then creates a table of all the loci for all traits and puts them together.
Then looks at the overlapps between the loci and creates a macro loci that do not overlap among them and assigns them a unique number (column pan_loci).
The function requires the sumstatst  have the columns: SNP, CHR,BP, A2 (EFFECT_ALLELE), A1 (NON_EFFECT_ALLELE), BETA,SE, P.

```{r}


locus_lister <- function(my_paths, gwas_names) {
  
      require(data.table)
      require(dplyr)
      require(GenomicRanges)
      
      SNPs <- list()
      list_of_files <- list()
      
      for( i in c(1:length(my_paths))){
        
            #load the sumstats and put them into a list
            list_of_files[[i]] <- fread(my_paths[[i]], data.table = F)
            SNPs[[i]]<- list_of_files[[i]]$SNP
            
      }
      
      shared_SNPs <- Reduce(intersect, SNPs) #compute the SNPs that are present in all the summary stats
      names(list_of_files) <- unlist(gwas_names)
      loci <- list()
      
      for(i in c(1:length(my_paths))){
        
            to_take <- gwas_names[[i]]
            sstats <- list_of_files[[to_take]]
            colnames(sstats) <- toupper(colnames(sstats))
            
            sstats <- select(sstats, c(SNP, CHR,BP, A2, A1, BETA,SE, P))
            sstats <- sstats[sstats$SNP %in% shared_SNPs,] #select only the SNPs that are shared among all the traits
            loci[[i]] <-  locus.breaker(sstats)  #run locus breaker
            loci[[i]]$trait <- rep(to_take, nrow(loci[[i]])) #name the loci
            
      }
      
      all_loci <- do.call(rbind, loci) #create the list of all the loci
      
      pan_loci <- reduce(GRanges(seqnames = all_loci$chr, IRanges(as.numeric(all_loci$start), as.numeric(all_loci$end)))) 
      pan_loci_non_reduced <- GRanges(seqnames = all_loci$chr, IRanges(as.numeric(all_loci$start), as.numeric(all_loci$end)))
      
      overlapping <- findOverlaps(pan_loci_non_reduced, pan_loci) #find overlaps between all the loci and use it as an index of the unique non overlapping loci
      all_loci$pan_locus <- rep(0, nrow(all_loci)) #allocate the column 
      all_loci[overlapping@from,]$pan_locus <- overlapping@to  #assigning the number as index of which macro loci is overlapping 
      all_loci$pan_locus_name <- rep(0, nrow(all_loci))
      
      #assign a name refereed to the position of each pan_locus
      for(k in 1:length(unique(all_loci$pan_locus))){
        all_loci[all_loci$pan_locus==k, ]$pan_locus_name <- paste0(all_loci[all_loci$pan_locus==k, ]$chr,'_',  min(all_loci[all_loci$pan_locus==k, ]$start), '_',  max(all_loci[all_loci$pan_locus==k, ]$end))
      }
      
      rownames(all_loci) <- NULL
      return(all_loci)
  
}

```



### run locus_lister on all the gwas (traits and factors)

```{r message=TRUE, warning=TRUE, paged.print=TRUE}
#regions all gwas

my_paths <- list("outputs/rev_1/05_munge/munged/asthma_munged_build37.txt",
                "outputs/rev_1/05_munge/munged/cd_munged_build37.txt",
                "outputs/rev_1/05_munge/munged/derma_munged_build37.txt",
              ###
                "outputs/rev_1/05_munge/munged/f1_munged_build37.txt",
                "outputs/rev_1/05_munge/munged/f2_munged_build37.txt",
                "outputs/rev_1/05_munge/munged/f3_munged_build37.txt",
              ###
                "outputs/rev_1/05_munge/munged/jia_munged_build37.txt",
                "outputs/rev_1/05_munge/munged/psc_munged_build37.txt",
                "outputs/rev_1/05_munge/munged/ra_munged_build37.txt",
              ###
                "outputs/rev_1/05_munge/munged/sle_munged_build37.txt",
                "outputs/rev_1/05_munge/munged/t1d_munged_build37.txt",
                "outputs/rev_1/05_munge/munged/uc_munged_build37.txt"
              )



gwas_names <- list('asthma','cd','derma',
                   'f1','f2','f3',
                   'jia','psc', 'ra',
                   'sle', 't1d', 'uc')


all_regions <- locus_lister(my_paths, gwas_names)
fwrite(all_regions, 'outputs/rev_1/06_genomic_regions_all_traits_munged/loci_all_traits.txt', sep='\t', col.names = T, row.names = F)

```

## Search for ovelapping genomic regions between factors after munge and locus_lister

```{r}


all_regions <- fread('outputs/rev_1/06_genomic_regions_all_traits_munged/loci_all_traits.txt', data.table=F)

f_ranges <- list()
for(i in c('f1', 'f2', 'f3')){
  
  f_ranges[[i]] <- GRanges(seqnames = all_regions[all_regions$trait==i,]$chr, IRanges(as.numeric(all_regions[all_regions$trait==i,]$start),as.numeric(all_regions[all_regions$trait==i,]$end)))
  
}

#plot venndiagram

venn_f1_f2_f3_loci<- makeVennDiagram(Peaks=f_ranges,connectedPeaks = 'keepAll',
                                     NameOfPeaks=c("F1", "F2", "F3"),
                                     imagetype="png" ,
                                     height = 480 , 
                                     width = 480 , 
                                     resolution = 300,
                                     compression = "lzw",
                                     lwd = 1,
                                     cex = 1.5,
                                     cat.cex = 2,
                                     cat.default.pos = "outer",
                                     cat.pos = c(-27, 27, 135),
                                     cat.dist = c(0.055, 0.055, 0.085),
                                     cat.fontfamily = "sans",
                                     rotation = 1
)



#----- plot UpSet plot ---------------------------------------------------------

#function for obtaining a combination matrix for plotting and UpSet plot from Granges names
lists_forupset <- function(peaks, traits){ 
      #allocate lists
      require(ComplexHeatmap)
      require(ChIPpeakAnno)
      require(stringr)
      
      ovl_f <- findOverlapsOfPeaks(peaks)
      
      un <- list()
      final <- list(list())
      sh <- list()
      #create the lists and put it the elements that are unique for each trait
      for( i in 1:length(traits)){
        
            tt<-traits[i]
            filt <- str_starts(names(ovl_f$uniquePeaks@ranges),tt, negate = FALSE)
            un[[tt]] <- names(ovl_f$uniquePeaks@ranges)[filt]
            
      }
      
      #put into the lists the element that are shared among the lists
      for(k in 1:length(ovl_f$mergedPeaks$peakNames)){
        
            chek_in <- traits %in% substring(ovl_f$mergedPeaks$peakNames[[k]], first = 1, last = 2) #check if there is f1, f2 or f3
            
            for(u in 1:length(chek_in)){
                  
                  tt<-traits[u]
                  
                  if(chek_in[u]==T){
                    
                      un[[tt]][[length(un[[tt]])+1]] <- paste0(ovl_f$mergedPeaks$peakNames[[k]], collapse = '-')
                    
                  }
              
            }
      }
      
      #generate the output
      
      cm <- make_comb_mat(un, mode = 'distinct')
      return(cm)
  
}


#plot

cm <- lists_forupset(f_ranges, traits = c('f1', 'f2', 'f3')) 


UpSet(cm, set_order = c("f1", "f2", "f3"), comb_order = order(comb_size(cm), decreasing = F),
      comb_col = c(brewer.pal(8, 'Set2')[c(7,6,6,6)],brewer.pal(12, 'Paired')[c(2,4,6)]),
      top_annotation = upset_top_annotation(cm, add_numbers = TRUE, height = unit(6, "cm")),
      left_annotation = upset_left_annotation(cm, add_numbers = TRUE, width = unit(3,'cm'), gp = gpar(fill = brewer.pal(5, 'Greys')[1])  ),
      row_title = "Factor", 
      column_title = " Physical intersection of factor loci"
)

```

## Exclude HLA region

We did not colocalize genomic regions that were overlapping the region 25,000,000 35,000,000 on chromosome 6. 

```{r paged.print=FALSE}
regions <- fread('outputs/rev_1/06_genomic_regions_all_traits_munged/loci_all_traits.txt', data.table = F)

#separate the pan_locus start and end column  
regions <- regions %>% tidyr::separate(col= pan_locus_name, sep = '_|_', into = c('CHR', 'START', 'END'))

#create a Granges object of the pan loci and 
new_rang <- GRanges(seqnames = regions$chr, IRanges(start = as.numeric(regions$START), as.numeric(regions$END)))
new_rang@elementMetadata$locus_name <- regions$pan_locus

#find the overlaps with chr 6, 25,000,000 - 35,000,000
aa <- IRanges::findOverlaps(new_rang, GRanges(seqnames = 6, IRanges(25000000, 35000000)) )

#return the hla loci
hla <- sort(unique(new_rang[aa@from,]@elementMetadata$locus_name))
regions[regions$pan_locus %in% hla,]
```












